### PR TITLE
RFC for KSN API

### PR DESCRIPTION
**Pull request by MeirShpilraien**
[Rendered](https://github.com/RedisGears/RedisGears/pull/896/files?short_path=5ad52b9#diff-5ad52b926da33590d90719661d104664846f23405a5d794315c95c1631847c3c)

### CODE REVIEW

File: rfcs/3-ksn-api.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+Inside the KSN callback, it is possible to perform any read operation but **it is forbidden to perform any write operations**. The reason for this is describe in details [here](https://github.com/redis/redis/pull/11199).
+
+In order to still perform a write operations as a reaction to KSN. Redis provides a new API called post notification job. This new API allows (inside the notification callback) to register a job that will be invoked when the following conditions holds:
+
+1. It is safe to perform any write operation.
+2. The job will be called atomically along side the key space notification that triggered it.
+
+This document describe options of exposing the described API to the JS user.
+
+## Option 2 - KSN only
```
Comment (thomascaudron) [inline rfcs/3-ksn-api.md:24]: option 1

File: rfcs/3-ksn-api.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    // this callback will be called after all the KSN finished.
+    // it is safe to write and read.
+    // the callback has access to the notification name and key name.
+    // The callback has not access to the data at the time the notification happened.
+    let key_name = data.key;
+    client.call('set', 'x', '1'); // performing write command
+});
+
+// Advance API for advance users
+redis.register_ksn_consumer("consumer", "", function(client, data){
```
Comment (K-Jo) [inline rfcs/3-ksn-api.md:98]: can we make the `ksn` in the function name an optional parameter of the function, so it's the same API?

File: rfcs/3-ksn-api.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    // This callback will be called directly on KSN.
+    // It is impossible to perform writes on this callback
+    // the callback can register a post notification job
+    // to write data atomically with the notification.
+    client.register_post_notification_job(function(client) {
```
Comment (K-Jo) [inline rfcs/3-ksn-api.md:103]: shouldn't we allow to pass some data here?

**Comment by thomascaudron** [comment]
**Can we split the creation of the trigger from the callbacks?**

```js
const trigger = redis.register_notification_consumer("consumer", "keys:");
const sharedData = {};

// This callback will be called directly on KSN.
// It is impossible to perform writes on this callback
trigger.onChange(client, data) {
    sharedData.propertyA = client.call("hget", data.key, "propertyA");
    sharedData.propertyA = ++;
};

// this callback will be called after all the KSN finished.
// it is safe to write and read.
// the callback has access to the notification name and key name.
// The callback has not access to the data at the time the notification happened.
trigger.postChange(client, data){
    client.call("hset", data.key, "propertyA", sharedData.propertyA);
};
```

It makes it clear when it will be executed without going into the details of the KSN.

**Where will be validate that no writes are happening?**
Can we return an error when the library is being loaded or throw an error when it is executed?

File: rfcs/3-ksn-api.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+## Option 4 - Simple API for regular users + advance API for advance users
+
+This option is a combination on 2 and 4. We will implement a simple API for most users which will look like option 2. For advance users we will implement option 4.
+
+Example:
+
+```js
+// Simple API for most users
+redis.register_notifications_consumer("consumer", "", function(client, data){
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply rfcs/3-ksn-api.md:88]: You can call it whatever you want, its a function argument.

File: rfcs/3-ksn-api.md (hunk 2 above)
Comment (MeirShpilraien) [reply rfcs/3-ksn-api.md:98]: Its a key space notification indicating key space changed. Just trigger does give you any context of why it has been triggered.

Comment (K-Jo) [reply rfcs/3-ksn-api.md:98]: register_crud_trigger?

**Pull request closed without merging**